docs.feldera.com/docs/operations/memory.md

```diff
@@ -0,0 +1,122 @@
+  connectors report their additional memory use as
+  [`input_connector_extra_memory_bytes`].
+
+  [`max_queued_records`]: /connectors#max_queued_records
+  [Kafka input connector]: /connectors/sources/kafka.md
+  [`records_input_buffered`]: metrics.md#records_input_buffered
+  [`records_input_buffered_bytes`]: metrics.md#records_input_buffered_bytes
+  [`input_connector_buffered_records`]: metrics.md#input_connector_buffered_records
+  [`input_connector_buffered_records_bytes`]: metrics.md#input_connector_buffered_records_bytes
+  [`input_connector_extra_memory_bytes`]: metrics.md#input_connector_extra_memory_bytes
+
+- **Output** records produced by the circuit, but not yet processed by
+  connectors.
+
+  As for input connectors, [`max_queued_records`] limits the maximum
+  number of records buffered.  This should be large enough to avoid
+  stalling the pipeline, but if the value is too high then a large
+  number of records can use excessive memory.  The default is
+  1,000,000.
+
+  The number of batches of buffered output records is exposed as
+  [`output_buffered_batches`], but since each batch contains a
+  variable number of records, this does not directly relate to memory
+  use.  The number of records buffered by individual connectors are
+  exposed as [`output_connector_buffered_records`].
+
+  [`output_buffered_batches`]: metrics.md#output_buffered_batches
+  [`output_connector_buffered_records`]: metrics.md#output_connector_buffered_records
+
+- **In-flight batches**.  As the pipeline executes a step, it passes
+  data batches from one operator to another.  Several data batches can
+  be in flight at any given time during a step.  The number and size
+  of these batches depends on the data, the number of records passed
+  in by the input connectors, the SQL program, the query plan
+  generated by Feldera's SQL compiler, and how Feldera schedules
+  execution of the query plan.  Most data batches only live as long as
+  it takes for them to be processed by an operator; they may be
+  transformed into part of the operator's output, or be added to an
+  index (see below), or be passed to an output connector, or simply be
+  discarded.
+
+  Since in-flight batches are transient, they do not usually become a
+  memory problem.  If they do, one may reduce [`max_batch_size`] for
+  input connectors, limiting the size of input batches.  Another
+  approach is to set `min_step_storage_bytes` in `storage` in the
+  pipeline [Runtime configuration], to force in-flight batches to
+  storage, although this is likely to reduce performance.
+
+  [`max_batch_size`]: /connectors#max_batch_size
+
+- **Index batches in memory**.  The pipeline primarily processes
+  records in memory.  As it does, it adds batches of records to its
+  in-memory indexes and then merges them into larger batches in the
+  background.  When merged batches grow large enough, they are flushed
+  to storage.  The minimum size to flush a batch to disk is
```

Comment:
this is another place where I think we should be a bit careful with wording. People are often confused about the whole compute in memory vs spill to disk vs keep everything in storage model. Saying that we process records in memory can give them the wrong idea (even though technically it's true).

I would just say that Feldera keeps state in on-disk indexes organized as LSM trees. The most recently added layers of the tree are temporarily cached in memory as a performance optimization. etc

Reply:
How's this?

```
- **Index batches in memory**.  The pipeline initially adds batches of
  records to its in-memory indexes and then merges them into larger
  batches in the background.  When a merged batch is large enough,
  it is written to storage.  The minimum size to write a batch to
  disk is configurable as `min_storage_bytes` under `storage` in the
  pipeline [Runtime configuration].  The default is 1 MiB, which is
  usually a good choice.  Configuring a smaller value may save memory
  but at a performance cost.
```